lib/nerves/release.ex

```diff
@@ -0,0 +1,199 @@
+defmodule Nerves.Release do
+  # No leading '/' here since this is passed to mksquashfs and it
+  # doesn't like the leading slash.
+  @target_release_path "srv/erlang"
+
+  def init(%{options: options} = release) do
+    opts = Keyword.merge(options, release_opts())
+
+    release = %{
+      release
+      | options: opts,
+        boot_scripts: [],
+        steps: release.steps ++ [&Nerves.Release.finalize/1]
+    }
+
+    if Code.ensure_loaded?(Shoehorn.Release) do
+      apply(Shoehorn.Release, :init, [release])
+    else
+      release
+    end
+  end
+
+  def finalize(release) do
+    bootfile_path = Path.join([release.version_path, bootfile()])
+
+    case File.read(bootfile_path) do
+      {:ok, bootfile} ->
+        Nerves.Release.write_rootfs_priorities(release.applications, release.path, bootfile)
+
+      _ ->
+        Nerves.Utils.Shell.warn("""
+          Unable to load bootfile: #{inspect(bootfile_path)}
+          Skipping rootfs priority file generation
+        """)
+    end
+
+    release
+  end
+
+  def bootfile() do
+    Application.get_env(:nerves, :firmware)[:bootfile] || "shoehorn.boot"
+  end
+
+  def erts() do
+    if Nerves.Env.loaded?() do
+      System.get_env("ERTS_DIR")
+    else
+      true
+    end
+  end
+
+  def write_rootfs_priorities(applications, host_release_path, bootfile) do
+    # Distillery support
+    applications = normalize_applications(applications)
+
+    target_release_path = @target_release_path
+
+    {:script, _, boot_script} = :erlang.binary_to_term(bootfile)
+
+    target_beam_files = target_beam_files(boot_script, host_release_path, target_release_path)
+    target_app_files = target_app_files(applications, target_release_path)
+    target_priv_dirs = target_priv_dirs(applications, target_release_path)
+
+    priorities =
+      (target_beam_files ++ target_app_files ++ target_priv_dirs)
+      |> List.flatten()
+      |> Enum.zip(32_000..1_000)
+      |> Enum.map(fn {file, priority} ->
+        file <> " " <> to_string(priority)
+      end)
+      |> Enum.join("\n")
+
+    build_path = Path.join([Mix.Project.build_path(), "nerves"])
+    File.mkdir_p(build_path)
+
+    Path.join(build_path, "rootfs.priorities")
+    |> File.write(priorities)
+  end
+
+  defp target_beam_files(boot_script, host_release_path, target_release_path) do
```

Comment:
Not a blocker.

There is a lot of complexity in this function - it might be a good target for refactoring.

Reply:
Yeah, this one was rough. I think a lot of the added complexity is that we are between release tools and have to support both distillery and elixir 1.9+. Ill take a pass through and see if I can make it easier or add more comments.